{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1d50b3e7",
   "metadata": {
    "papermill": {
     "duration": 0.004478,
     "end_time": "2025-06-18T07:12:26.950570",
     "exception": false,
     "start_time": "2025-06-18T07:12:26.946092",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "This hands-on exercise demonstrates how to train and use the GEARS model (Roohani Y, et al. Nature Biotechnology, 2024) with a gene perturbation dataset. You can plug the LCM into the Perturbation task by changing the gene embedding from GEARS initial embedding to LCMs' gene representation. \n",
    "\n",
    "Due to computational resources, we won't be calling the LCMs here to get cell representations. If you are interested, you can try it on your own after class.\n",
    "\n",
    "We will first specifically show the content of the perturbation data, then construct the GEARS model and train it, and finally use the trained model for perturbation prediction and evaluate the effectiveness of the prediction."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1875dfc4",
   "metadata": {
    "papermill": {
     "duration": 0.003358,
     "end_time": "2025-06-18T07:12:26.958069",
     "exception": false,
     "start_time": "2025-06-18T07:12:26.954711",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Import necessary library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "562124e4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:12:26.965577Z",
     "iopub.status.busy": "2025-06-18T07:12:26.965358Z",
     "iopub.status.idle": "2025-06-18T07:13:18.637054Z",
     "shell.execute_reply": "2025-06-18T07:13:18.636455Z"
    },
    "papermill": {
     "duration": 51.677168,
     "end_time": "2025-06-18T07:13:18.638550",
     "exception": false,
     "start_time": "2025-06-18T07:12:26.961382",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import torch\n",
    "import pickle\n",
    "\n",
    "import sys\n",
    "sys.path.append('/kaggle/input/gears-tutorial/pytorch/default/1/GEARS')\n",
    "from gears import GEARS\n",
    "from gears.inference import evaluate\n",
    "from gears.model import GEARS_Model\n",
    "from gears.utils import get_similarity_network, GeneSimNetwork\n",
    "from copy import deepcopy\n",
    "                  \n",
    "from sklearn.metrics import r2_score\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "from sklearn.metrics import mean_squared_error as mse\n",
    "from sklearn.metrics import mean_absolute_error as mae"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8d388ba9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:13:18.646938Z",
     "iopub.status.busy": "2025-06-18T07:13:18.646404Z",
     "iopub.status.idle": "2025-06-18T07:13:18.825742Z",
     "shell.execute_reply": "2025-06-18T07:13:18.825158Z"
    },
    "papermill": {
     "duration": 0.184618,
     "end_time": "2025-06-18T07:13:18.826985",
     "exception": false,
     "start_time": "2025-06-18T07:13:18.642367",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "sys.path.append(\"/kaggle/input/pert_dataset/pytorch/default/1/Pert_Data/\")\n",
    "import v1\n",
    "from v1.utils import *\n",
    "from v1.dataloader import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30cf84ab",
   "metadata": {
    "papermill": {
     "duration": 0.003253,
     "end_time": "2025-06-18T07:13:18.833993",
     "exception": false,
     "start_time": "2025-06-18T07:13:18.830740",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Load perturbation dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6af847c",
   "metadata": {
    "papermill": {
     "duration": 0.003157,
     "end_time": "2025-06-18T07:13:18.840407",
     "exception": false,
     "start_time": "2025-06-18T07:13:18.837250",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "**Load and process data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d61b9303",
   "metadata": {
    "_kg_hide-output": true,
    "execution": {
     "iopub.execute_input": "2025-06-18T07:13:18.847878Z",
     "iopub.status.busy": "2025-06-18T07:13:18.847485Z",
     "iopub.status.idle": "2025-06-18T07:15:39.347544Z",
     "shell.execute_reply": "2025-06-18T07:15:39.346689Z"
    },
    "papermill": {
     "duration": 140.505152,
     "end_time": "2025-06-18T07:15:39.348748",
     "exception": false,
     "start_time": "2025-06-18T07:13:18.843596",
     "status": "completed"
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========== read file finished!\n",
      "retain_pert_num is: 196\n",
      "filtered pert num is:  41\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 236/236 [00:04<00:00, 48.26it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========== filter perturbation finished!\n",
      "len of exclude_var_list is 4\n",
      "len of pert_gene_list is 289\n",
      "len of final var_names is 1067\n",
      "========== get var genes finished!\n",
      "this is new version\n",
      "========== data split finished!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 195/195 [00:04<00:00, 47.93it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========== set control barcodes finished!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/48 [00:00<?, ?it/s]2025-06-18 07:14:42.840812: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
      "E0000 00:00:1750230883.023344      46 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "E0000 00:00:1750230883.083045      46 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "\n",
      "100%|██████████| 10/10 [00:00<00:00, 554.14it/s]\n",
      "  2%|▏         | 1/48 [00:27<21:15, 27.14s/it]\n",
      "100%|██████████| 6/6 [00:00<00:00, 129.66it/s]\n",
      "  4%|▍         | 2/48 [00:27<08:44, 11.40s/it]\n",
      "100%|██████████| 6/6 [00:00<00:00, 101.37it/s]\n",
      "  6%|▋         | 3/48 [00:27<04:46,  6.36s/it]\n",
      "100%|██████████| 6/6 [00:00<00:00, 73.72it/s]\n",
      "  8%|▊         | 4/48 [00:28<02:57,  4.02s/it]\n",
      "100%|██████████| 6/6 [00:00<00:00, 78.98it/s]\n",
      " 10%|█         | 5/48 [00:28<01:56,  2.72s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C19orf26 not in var!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 6/6 [00:00<00:00, 88.70it/s]\n",
      " 12%|█▎        | 6/48 [00:29<01:20,  1.93s/it]\n",
      "100%|██████████| 6/6 [00:00<00:00, 114.66it/s]\n",
      " 15%|█▍        | 7/48 [00:29<00:58,  1.42s/it]\n",
      "100%|██████████| 6/6 [00:00<00:00, 200.91it/s]\n",
      " 17%|█▋        | 8/48 [00:29<00:42,  1.06s/it]\n",
      "100%|██████████| 6/6 [00:00<00:00, 146.03it/s]\n",
      " 19%|█▉        | 9/48 [00:30<00:32,  1.20it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 104.77it/s]\n",
      " 21%|██        | 10/48 [00:30<00:26,  1.46it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 144.28it/s]\n",
      " 23%|██▎       | 11/48 [00:30<00:21,  1.72it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 87.73it/s]\n",
      " 25%|██▌       | 12/48 [00:31<00:18,  1.90it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 121.55it/s]\n",
      " 27%|██▋       | 13/48 [00:31<00:16,  2.09it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 97.77it/s]\n",
      " 29%|██▉       | 14/48 [00:31<00:15,  2.23it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 95.87it/s]\n",
      " 31%|███▏      | 15/48 [00:32<00:14,  2.32it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 80.45it/s]\n",
      " 33%|███▎      | 16/48 [00:32<00:13,  2.34it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 141.95it/s]\n",
      " 35%|███▌      | 17/48 [00:33<00:12,  2.52it/s]\n",
      "  0%|          | 0/6 [00:00<?, ?it/s]\u001b[A\n",
      "100%|██████████| 6/6 [00:00<00:00, 55.43it/s]\n",
      " 38%|███▊      | 18/48 [00:33<00:13,  2.29it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 110.65it/s]\n",
      " 40%|███▉      | 19/48 [00:34<00:12,  2.37it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 353.46it/s]\n",
      " 42%|████▏     | 20/48 [00:34<00:10,  2.79it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 79.04it/s]\n",
      " 44%|████▍     | 21/48 [00:34<00:10,  2.64it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 70.57it/s]\n",
      " 46%|████▌     | 22/48 [00:35<00:10,  2.50it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 277.82it/s]\n",
      " 48%|████▊     | 23/48 [00:35<00:09,  2.76it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 331.68it/s]\n",
      " 50%|█████     | 24/48 [00:35<00:07,  3.02it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 134.46it/s]\n",
      " 52%|█████▏    | 25/48 [00:35<00:07,  2.96it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 124.68it/s]\n",
      " 54%|█████▍    | 26/48 [00:36<00:07,  2.97it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 130.10it/s]\n",
      " 56%|█████▋    | 27/48 [00:36<00:07,  2.95it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 123.93it/s]\n",
      " 58%|█████▊    | 28/48 [00:37<00:06,  2.94it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 113.51it/s]\n",
      " 60%|██████    | 29/48 [00:37<00:06,  2.91it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 115.10it/s]\n",
      " 62%|██████▎   | 30/48 [00:37<00:06,  2.91it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 104.35it/s]\n",
      " 65%|██████▍   | 31/48 [00:38<00:05,  2.89it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 213.32it/s]\n",
      " 67%|██████▋   | 32/48 [00:38<00:05,  3.01it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 232.57it/s]\n",
      " 69%|██████▉   | 33/48 [00:38<00:05,  2.84it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 171.53it/s]\n",
      " 71%|███████   | 34/48 [00:39<00:05,  2.76it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 229.77it/s]\n",
      " 73%|███████▎  | 35/48 [00:39<00:04,  2.92it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 108.25it/s]\n",
      " 75%|███████▌  | 36/48 [00:39<00:04,  2.86it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 103.41it/s]\n",
      " 77%|███████▋  | 37/48 [00:40<00:03,  2.80it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 70.32it/s]\n",
      " 79%|███████▉  | 38/48 [00:40<00:03,  2.61it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 148.19it/s]\n",
      " 81%|████████▏ | 39/48 [00:40<00:03,  2.75it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 395.43it/s]\n",
      " 83%|████████▎ | 40/48 [00:41<00:02,  3.12it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 216.22it/s]\n",
      " 85%|████████▌ | 41/48 [00:41<00:02,  2.78it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 83.70it/s]\n",
      " 88%|████████▊ | 42/48 [00:42<00:02,  2.68it/s]\n",
      "  0%|          | 0/6 [00:00<?, ?it/s]\u001b[A\n",
      "100%|██████████| 6/6 [00:00<00:00, 35.70it/s]\n",
      " 90%|████████▉ | 43/48 [00:42<00:02,  2.16it/s]\n",
      "  0%|          | 0/6 [00:00<?, ?it/s]\u001b[A\n",
      "100%|██████████| 6/6 [00:00<00:00, 58.71it/s]\n",
      " 92%|█████████▏| 44/48 [00:43<00:02,  1.98it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 211.36it/s]\n",
      " 94%|█████████▍| 45/48 [00:44<00:02,  1.46it/s]\n",
      "  0%|          | 0/6 [00:00<?, ?it/s]\u001b[A\n",
      " 50%|█████     | 3/6 [00:00<00:00, 24.54it/s]\u001b[A\n",
      "100%|██████████| 6/6 [00:00<00:00, 22.04it/s]\n",
      " 96%|█████████▌| 46/48 [00:45<00:01,  1.31it/s]\n",
      "  0%|          | 0/6 [00:00<?, ?it/s]\u001b[A\n",
      "100%|██████████| 6/6 [00:00<00:00, 52.64it/s]\n",
      " 98%|█████████▊| 47/48 [00:45<00:00,  1.42it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 92.69it/s]\n",
      "100%|██████████| 48/48 [00:46<00:00,  1.04it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========== HOXC13 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== FOXL2 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== MAP7D1 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== ELMSAN1 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== MEIS1 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== HOXB9 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== SAMD1 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== SPI1 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== DLX2 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== IGDCC3 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== UBASH3A | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== IKZF3 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== MAPK1 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== CNN1 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== FOXF1 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== ISL2 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== UBASH3B | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== ZBTB1 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== TGFBR2; IGDCC3 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== MAP2K6 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== OSR2 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== IER5L | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== CDKN1A | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== KIF2C | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== CBFA2T3 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== TGFBR2 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== CLDN6 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== TBX2 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== TBX3 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== PRTG | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== PTPN12 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== CEBPA | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== CDKN1B | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== PTPN9 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== FEV | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== ZBTB25 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== SET | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== COL2A1 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== STIL | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== SNAI1 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== FOXA1 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== BAK1 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n",
      "========== ETS2 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== CEBPB | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== KLF1 | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== CEBPE | lymphoblasts\n",
      "filter_sgRNA_list is: []\n",
      "========== RUNX1T1 | lymphoblasts\n",
      "max_eids < 10, no need to filter sgRNA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/195 [00:00<?, ?it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  1%|          | 1/195 [00:06<21:14,  6.57s/it]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  2%|▏         | 3/195 [00:06<05:38,  1.76s/it]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  3%|▎         | 5/195 [00:06<02:51,  1.11it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  3%|▎         | 6/195 [00:07<02:11,  1.44it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  4%|▎         | 7/195 [00:07<01:40,  1.87it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  4%|▍         | 8/195 [00:07<01:16,  2.43it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  5%|▌         | 10/195 [00:07<00:50,  3.69it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  6%|▌         | 11/195 [00:07<00:44,  4.16it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  6%|▌         | 12/195 [00:07<00:38,  4.79it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  7%|▋         | 14/195 [00:07<00:28,  6.40it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  8%|▊         | 16/195 [00:08<00:22,  7.86it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "  9%|▉         | 18/195 [00:08<00:23,  7.50it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 10%|█         | 20/195 [00:08<00:20,  8.65it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 11%|█▏        | 22/195 [00:08<00:17,  9.68it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 12%|█▏        | 24/195 [00:08<00:16, 10.34it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 13%|█▎        | 26/195 [00:08<00:15, 10.86it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 14%|█▍        | 28/195 [00:09<00:14, 11.40it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 15%|█▌        | 30/195 [00:09<00:15, 10.64it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 16%|█▋        | 32/195 [00:09<00:14, 11.44it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 17%|█▋        | 34/195 [00:09<00:16,  9.78it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 18%|█▊        | 36/195 [00:10<00:17,  9.31it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 19%|█▉        | 38/195 [00:10<00:16,  9.72it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 21%|██        | 40/195 [00:10<00:17,  8.95it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 22%|██▏       | 42/195 [00:10<00:15,  9.70it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 23%|██▎       | 44/195 [00:10<00:15,  9.99it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 24%|██▎       | 46/195 [00:10<00:13, 10.73it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 25%|██▍       | 48/195 [00:11<00:13, 11.25it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 26%|██▌       | 50/195 [00:11<00:12, 11.96it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 27%|██▋       | 52/195 [00:11<00:12, 11.69it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 28%|██▊       | 54/195 [00:11<00:11, 12.36it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 29%|██▊       | 56/195 [00:11<00:11, 11.88it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 30%|██▉       | 58/195 [00:11<00:12, 11.35it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 31%|███       | 60/195 [00:12<00:11, 11.26it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 32%|███▏      | 62/195 [00:12<00:12, 11.02it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 33%|███▎      | 64/195 [00:12<00:11, 11.51it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 34%|███▍      | 66/195 [00:12<00:10, 12.11it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 35%|███▍      | 68/195 [00:12<00:10, 11.76it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 36%|███▌      | 70/195 [00:12<00:10, 11.67it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 37%|███▋      | 72/195 [00:13<00:10, 12.01it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 38%|███▊      | 74/195 [00:13<00:10, 11.77it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 39%|███▉      | 76/195 [00:13<00:10, 11.83it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 40%|████      | 78/195 [00:13<00:09, 12.04it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 41%|████      | 80/195 [00:13<00:09, 11.88it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 42%|████▏     | 82/195 [00:14<00:10, 10.93it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 43%|████▎     | 84/195 [00:14<00:09, 11.22it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 44%|████▍     | 86/195 [00:14<00:09, 11.34it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 45%|████▌     | 88/195 [00:14<00:09, 11.17it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 46%|████▌     | 90/195 [00:14<00:10, 10.18it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 47%|████▋     | 92/195 [00:14<00:09, 10.39it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 48%|████▊     | 94/195 [00:15<00:10, 10.07it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 49%|████▉     | 96/195 [00:15<00:09,  9.91it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 50%|█████     | 98/195 [00:15<00:09, 10.54it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 51%|█████▏    | 100/195 [00:15<00:08, 11.25it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 52%|█████▏    | 102/195 [00:15<00:08, 10.72it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 53%|█████▎    | 104/195 [00:16<00:08, 10.41it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 54%|█████▍    | 106/195 [00:16<00:08, 10.76it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 55%|█████▌    | 108/195 [00:16<00:07, 11.09it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 56%|█████▋    | 110/195 [00:16<00:07, 11.50it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 57%|█████▋    | 112/195 [00:16<00:07, 11.65it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 58%|█████▊    | 114/195 [00:16<00:06, 12.29it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 59%|█████▉    | 116/195 [00:17<00:06, 12.29it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 61%|██████    | 118/195 [00:17<00:06, 12.31it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 62%|██████▏   | 120/195 [00:17<00:06, 12.47it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 63%|██████▎   | 122/195 [00:17<00:05, 12.49it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 64%|██████▎   | 124/195 [00:17<00:06, 11.64it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 65%|██████▍   | 126/195 [00:17<00:05, 12.19it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 66%|██████▌   | 128/195 [00:18<00:05, 12.17it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 67%|██████▋   | 130/195 [00:18<00:05, 12.39it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 68%|██████▊   | 132/195 [00:18<00:05, 12.29it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 69%|██████▊   | 134/195 [00:18<00:05, 12.15it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 70%|██████▉   | 136/195 [00:18<00:04, 12.32it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 71%|███████   | 138/195 [00:18<00:04, 12.12it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 72%|███████▏  | 140/195 [00:19<00:04, 12.00it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 73%|███████▎  | 142/195 [00:19<00:04, 12.18it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 74%|███████▍  | 144/195 [00:19<00:04, 11.76it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 75%|███████▍  | 146/195 [00:19<00:04, 11.60it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 76%|███████▌  | 148/195 [00:19<00:03, 12.12it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 77%|███████▋  | 150/195 [00:19<00:03, 11.85it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 78%|███████▊  | 152/195 [00:20<00:03, 11.56it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 79%|███████▉  | 154/195 [00:20<00:03, 11.99it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 80%|████████  | 156/195 [00:20<00:03, 11.94it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 81%|████████  | 158/195 [00:20<00:03, 11.91it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 82%|████████▏ | 160/195 [00:20<00:02, 11.81it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 83%|████████▎ | 162/195 [00:20<00:02, 11.62it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 84%|████████▍ | 164/195 [00:21<00:02, 12.23it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 85%|████████▌ | 166/195 [00:21<00:02, 12.70it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 86%|████████▌ | 168/195 [00:21<00:02, 13.03it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 87%|████████▋ | 170/195 [00:21<00:01, 12.71it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 88%|████████▊ | 172/195 [00:21<00:01, 12.92it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 89%|████████▉ | 174/195 [00:21<00:01, 12.44it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 90%|█████████ | 176/195 [00:22<00:01, 12.52it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 91%|█████████▏| 178/195 [00:22<00:01, 13.06it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 92%|█████████▏| 180/195 [00:22<00:01, 13.22it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 93%|█████████▎| 182/195 [00:22<00:00, 13.38it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 94%|█████████▍| 184/195 [00:22<00:00, 13.47it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 95%|█████████▌| 186/195 [00:22<00:00, 12.92it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 96%|█████████▋| 188/195 [00:22<00:00, 13.07it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 97%|█████████▋| 190/195 [00:23<00:00, 12.97it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 98%|█████████▊| 192/195 [00:23<00:00, 13.23it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      " 99%|█████████▉| 194/195 [00:23<00:00, 12.96it/s]... storing 'guide_id' as categorical\n",
      "... storing 'tissue_type' as categorical\n",
      "... storing 'cell_line' as categorical\n",
      "... storing 'disease' as categorical\n",
      "... storing 'perturbation_type' as categorical\n",
      "... storing 'celltype' as categorical\n",
      "... storing 'organism' as categorical\n",
      "... storing 'perturbation' as categorical\n",
      "... storing 'perturbation_new' as categorical\n",
      "... storing 'perturbation_type_new' as categorical\n",
      "... storing 'celltype_new' as categorical\n",
      "... storing 'sgRNA_new' as categorical\n",
      "... storing 'perturbation_group' as categorical\n",
      "... storing 'data_split' as categorical\n",
      "... storing 'retain' as categorical\n",
      "... storing 'control_barcode' as categorical\n",
      "... storing 'sgRNA_ID' as categorical\n",
      "... storing 'pert_sgRNA' as categorical\n",
      "100%|██████████| 195/195 [00:23<00:00,  8.32it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========== get de genes finished!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# - load pert_data and preprocess\n",
    "pert_data = Byte_Pert_Data(data_dir='/kaggle/input/example-data-pert/',prefix='NormanWeissman2019_filtered',) # NormanWeissman2019_filtered or XuCao2023\n",
    "pert_data.read_files()\n",
    "pert_data.filter_perturbation()   # filter perturbation with less cell number\n",
    "pert_data.get_and_process_adata(var_num=1000)    # process the data and obtain the higly variable genes\n",
    "pert_data.data_split(split_type=1)   # split the data into train and test set\n",
    "pert_data.set_control_barcode()   # set the control cell's barcode for each perturbed cell\n",
    "pert_data.filter_sgRNA()  # for each pert, get the sgRNA num\n",
    "pert_data.get_de_genes()  # calculate the DE genes for each perturbed cell; used for evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8bb91d38",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:15:39.705543Z",
     "iopub.status.busy": "2025-06-18T07:15:39.704412Z",
     "iopub.status.idle": "2025-06-18T07:15:54.570603Z",
     "shell.execute_reply": "2025-06-18T07:15:54.569703Z"
    },
    "papermill": {
     "duration": 15.019386,
     "end_time": "2025-06-18T07:15:54.571863",
     "exception": false,
     "start_time": "2025-06-18T07:15:39.552477",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found local copy...\n",
      "Found local copy...\n",
      " 68%|██████▊   | 92/136 [00:03<00:01, 28.96it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LYL1; IER5L | lymphoblasts not in pert_names\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 136/136 [00:05<00:00, 23.33it/s]\n",
      "100%|██████████| 39/39 [00:01<00:00, 36.74it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KIAA1804 | lymphoblasts not in pert_names\n",
      "IER5L | lymphoblasts not in pert_names\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 37.22it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========== get Data_gears finished!\n",
      "add adata finished\n",
      "add condition finished\n",
      "add set2conditions finished\n"
     ]
    }
   ],
   "source": [
    "fix_seed(2024)\n",
    "\n",
    "# - get go genes; special set for GEARS; these are used for constructing the go graph of GEARS\n",
    "pert_data.get_gene2go()   # get gene2go dict: {'gene1': [go1, go2, ...], 'gene2': [go1, go3, ...], ...}\n",
    "pert_data.set_pert_genes()  # get the list of genes that can be perturbed to be included in perturbation graph\n",
    "\n",
    "# - transform dataset into gears required format\n",
    "pert_data.get_Data_gears(num_de_genes = pert_data.num_de_genes,\n",
    "                        dataset_name = ['train', 'test', 'val'],\n",
    "                        add_control = False)\n",
    "# - add necessary elements for gears\n",
    "pert_data.modify_gears()\n",
    "\n",
    "# - get dataloader\n",
    "trainloader, testloader, valloader = pert_data.get_dataloader(mode='all')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1d5031b",
   "metadata": {
    "papermill": {
     "duration": 0.150068,
     "end_time": "2025-06-18T07:15:54.878864",
     "exception": false,
     "start_time": "2025-06-18T07:15:54.728796",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "**Dataset details**\n",
    "\n",
    "This dataset has 96852 cells and each of them has 1167 genes. It has 170 different perturbations. We have pre-processed the data and calculated the differential genes. These differential genes can be considered as significantly changed genes after perturbation and will be used to assess the effect of prediction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "354f6128",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:15:55.183524Z",
     "iopub.status.busy": "2025-06-18T07:15:55.183243Z",
     "iopub.status.idle": "2025-06-18T07:15:55.188995Z",
     "shell.execute_reply": "2025-06-18T07:15:55.188440Z"
    },
    "papermill": {
     "duration": 0.160466,
     "end_time": "2025-06-18T07:15:55.190061",
     "exception": false,
     "start_time": "2025-06-18T07:15:55.029595",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 106022 × 1067\n",
       "    obs: 'guide_id', 'read_count', 'UMI_count', 'coverage', 'gemgroup', 'good_coverage', 'number_of_cells', 'tissue_type', 'cell_line', 'cancer', 'disease', 'perturbation_type', 'celltype', 'organism', 'perturbation', 'nperts', 'ngenes', 'ncounts', 'percent_mito', 'percent_ribo', 'perturbation_new', 'perturbation_type_new', 'nperts_new', 'celltype_new', 'sgRNA_new', 'perturbation_group', 'data_split', 'retain', 'n_genes', 'n_counts_all', 'control_barcode', 'sgRNA_ID', 'pert_sgRNA', 'condition', 'condition_name'\n",
       "    var: 'ensemble_id', 'ncounts', 'ncells', 'gene_name'\n",
       "    uns: 'rank_genes_groups', 'pvals', 'pvals_adj', 'scores', 'logfoldchanges', 'top_non_dropout_de_20', 'non_dropout_gene_idx', 'non_zeros_gene_idx', 'top_non_zero_de_20', 'rank_genes_groups_cov_all'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pert_data.adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0b988a31",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:15:55.538377Z",
     "iopub.status.busy": "2025-06-18T07:15:55.537593Z",
     "iopub.status.idle": "2025-06-18T07:15:55.560073Z",
     "shell.execute_reply": "2025-06-18T07:15:55.559503Z"
    },
    "papermill": {
     "duration": 0.219548,
     "end_time": "2025-06-18T07:15:55.561169",
     "exception": false,
     "start_time": "2025-06-18T07:15:55.341621",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['KMT2A', 'MAP2K3; ELMSAN1', 'UBASH3B; CNN1', 'CEBPE; RUNX1T1',\n",
       "       'ETS2; MAPK1', 'SET; CEBPE', 'EGR1', 'FOXO4', 'CNN1', 'CELF2',\n",
       "       'PTPN12; PTPN9', 'ZNF318', 'ETS2; MAP7D1', 'DUSP9', 'MEIS1',\n",
       "       'ELMSAN1', 'POU3F2; FOXL2', 'MAP7D1', 'FOXL2', 'ETS2; PRTG',\n",
       "       'CSRNP1', 'ATL1', 'TP73', 'RREB1', 'TBX2', 'CDKN1C', 'KLF1; CEBPA',\n",
       "       'FOSB', 'CBFA2T3', 'ETS2; IKZF3', 'RHOXF2; ZBTB25', 'KIAA1804',\n",
       "       'CDKN1A', 'MAPK1; PRTG', 'IER5L', 'CITED1', 'ZC3HAV1',\n",
       "       'PTPN12; SNAI1', 'CEBPB; OSR2'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pert_data.adata[pert_data.adata.obs['data_split']=='test'].obs['perturbation_new'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0e9e336a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:15:55.864609Z",
     "iopub.status.busy": "2025-06-18T07:15:55.864378Z",
     "iopub.status.idle": "2025-06-18T07:15:55.899594Z",
     "shell.execute_reply": "2025-06-18T07:15:55.898838Z"
    },
    "papermill": {
     "duration": 0.187151,
     "end_time": "2025-06-18T07:15:55.900799",
     "exception": false,
     "start_time": "2025-06-18T07:15:55.713648",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CEBPB', 'TGFBR2; IGDCC3', 'control', 'SNAI1', 'CNN1; MAPK1',\n",
       "       'GLB1L2', 'OSR2', 'MAP2K6', 'LHX1; ELMSAN1', 'SET; KLF1', 'CBL',\n",
       "       'CEBPE', 'DUSP9; KLF1', 'S1PR2', 'CEBPE; PTPN12', 'UBASH3B; PTPN9',\n",
       "       'KLF1', 'TMSB4X; BAK1', 'SLC6A9', 'ETS2; IGDCC3', 'AHR; FEV',\n",
       "       'STIL', 'ARRDC3', 'TGFBR2; C19orf26', 'TSC22D1', 'COL2A1',\n",
       "       'FOXF1; HOXB9', 'FOXA1', 'BAK1', 'ETS2', 'CBL; CNN1', 'MAP2K3',\n",
       "       'SLC4A1', 'PTPN12; OSR2', 'COL1A1', 'LYL1', 'ZNF318; FOXL2',\n",
       "       'TMSB4X', 'KLF1; CLDN6', 'MAPK1; TGFBR2', 'ISL2', 'FOXA1; HOXB9',\n",
       "       'ZBTB10; PTPN12', 'FOXF1', 'BCL2L11', 'MAML2', 'AHR; KLF1', 'HK2',\n",
       "       'CEBPB; PTPN12', 'PRDM1', 'IKZF3', 'KLF1; TGFBR2', 'ETS2; CEBPE',\n",
       "       'FOSB; PTPN12', 'UBASH3A', 'MAP4K3', 'FOXA3', 'PTPN12; ZBTB25',\n",
       "       'FOXA3; FOXA1', 'DLX2', 'CBL; PTPN9', 'UBASH3B', 'SAMD1',\n",
       "       'TGFBR2; ETS2', 'SGK1; TBX3', 'HOXB9', 'LHX1', 'HOXC13',\n",
       "       'TBX3; TBX2', 'BCORL1', 'C19orf26', 'UBASH3B; PTPN12', 'HOXA13',\n",
       "       'IGDCC3; MAPK1', 'ETS2; CNN1', 'TGFBR2', 'FEV; ISL2', 'FEV',\n",
       "       'MAP2K3; IKZF3', 'PTPN9', 'BPGM; ZBTB1', 'NCL', 'SGK1; TBX2',\n",
       "       'DUSP9; PRTG', 'SGK1', 'MAP2K6; IKZF3', 'CKS1B', 'MAP4K5', 'CEBPA',\n",
       "       'UBASH3B; OSR2', 'LYL1; IER5L', 'RHOXF2; SET', 'MAP2K3; SLC38A2',\n",
       "       'MAP2K6; ELMSAN1', 'AHR', 'FOSB; OSR2', 'PRTG', 'PTPN12; UBASH3A',\n",
       "       'C3orf72', 'FOXA1; FOXF1', 'MAP2K3; MAP2K6', 'KLF1; BAK1', 'CLDN6',\n",
       "       'KLF1; FOXA1', 'ZBTB25', 'JUN', 'FOXA3; HOXB9', 'DUSP9; ETS2',\n",
       "       'BCL2L11; TGFBR2', 'DUSP9; MAPK1', 'CBL; PTPN12', 'CNNM4',\n",
       "       'CEBPE; KLF1', 'UBASH3B; ZBTB25', 'FOSB; UBASH3B', 'KIF2C',\n",
       "       'BPGM; SAMD1', 'MAPK1; IKZF3', 'RHOXF2', 'BPGM', 'ZC3HAV1; CEBPE',\n",
       "       'SAMD1; UBASH3B', 'IRF1', 'SLC38A2', 'PTPN1', 'FEV; MAP7D1',\n",
       "       'ARID1A', 'PTPN13', 'SAMD1; PTPN12', 'DUSP9; IGDCC3',\n",
       "       'FOSB; IKZF3', 'ZBTB1', 'SNAI1; UBASH3B', 'KLF1; COL2A1',\n",
       "       'IRF1; SET', 'HNF4A', 'POU3F2; CBFA2T3'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pert_data.adata[pert_data.adata.obs['data_split']=='train'].obs['perturbation_new'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ada7eda2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:15:56.203088Z",
     "iopub.status.busy": "2025-06-18T07:15:56.202839Z",
     "iopub.status.idle": "2025-06-18T07:15:56.208262Z",
     "shell.execute_reply": "2025-06-18T07:15:56.207620Z"
    },
    "papermill": {
     "duration": 0.156588,
     "end_time": "2025-06-18T07:15:56.209381",
     "exception": false,
     "start_time": "2025-06-18T07:15:56.052793",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACCTGAGAAACCAT             CEBPB\n",
       "AAACCTGAGAAAGTGG    TGFBR2; IGDCC3\n",
       "AAACCTGAGAAGAAGC           control\n",
       "AAACCTGAGAAGGTTT             SNAI1\n",
       "AAACCTGAGACATAAC       CNN1; MAPK1\n",
       "Name: perturbation_new, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# In the perturbation dataset, each cell has one (or more than one) perturbed gene\n",
    "pert_data.adata.obs['perturbation_new'][:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a065f500",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:15:56.518132Z",
     "iopub.status.busy": "2025-06-18T07:15:56.517586Z",
     "iopub.status.idle": "2025-06-18T07:15:56.522986Z",
     "shell.execute_reply": "2025-06-18T07:15:56.522296Z"
    },
    "papermill": {
     "duration": 0.163732,
     "end_time": "2025-06-18T07:15:56.524061",
     "exception": false,
     "start_time": "2025-06-18T07:15:56.360329",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACCTGAGAAACCAT    GGAATAACAAGGTTCT\n",
       "AAACCTGAGAAAGTGG    GGAATAACAAGGTTCT\n",
       "AAACCTGAGAAGAAGC                None\n",
       "AAACCTGAGAAGGTTT    GGAATAACAAGGTTCT\n",
       "AAACCTGAGACATAAC    GGAATAACAAGGTTCT\n",
       "Name: control_barcode, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# And also a control group cell, representing the cell before perturbation\n",
    "pert_data.adata.obs['control_barcode'][:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3cd77bd6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:15:56.876742Z",
     "iopub.status.busy": "2025-06-18T07:15:56.876453Z",
     "iopub.status.idle": "2025-06-18T07:15:56.881142Z",
     "shell.execute_reply": "2025-06-18T07:15:56.880261Z"
    },
    "papermill": {
     "duration": 0.158514,
     "end_time": "2025-06-18T07:15:56.882173",
     "exception": false,
     "start_time": "2025-06-18T07:15:56.723659",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CEBPB | lymphoblasts': array(['PLD3', 'SH3BGRL3', 'AIF1', 'LST1', 'FTL', 'RP11-301G19.1',\n",
      "       'TMSB10', 'TMSB4X', 'CFD', 'LGI2', 'ID1', 'ID3', 'GYPA', 'HBG2',\n",
      "       'GAL', 'MSRB1', 'CSF3R', 'ARPC1B', 'TYROBP', 'MYO1F'], dtype='<U13'), 'TGFBR2; IGDCC3 | lymphoblasts': array(['IGDCC3', 'ALAS2', 'HBZ', 'HBG2', 'RP11-301G19.1', 'GYPB',\n",
      "       'PRSS57', 'HBG1', 'BST2', 'CPEB4', 'FAM83A', 'GYPA', 'HBA1',\n",
      "       'IGFBP2', 'YBX1', 'MDK', 'TMSB10', 'GAL', 'AC079466.1', 'APOE'],\n",
      "      dtype='<U13'), 'SNAI1 | lymphoblasts': array(['TMSB10', 'HBG2', 'S100A13', 'LGALS1', 'HIST1H1C', 'RP11-717F1.1',\n",
      "       'CFD', 'RNASET2', 'GAL', 'HBZ', 'HIST1H2BJ', 'MDK', 'PRSS57',\n",
      "       'MT-ND2', 'ALAS2', 'BST2', 'YBX1', 'VIM', 'S100A11', 'NCL'],\n",
      "      dtype='<U12'), 'CNN1; MAPK1 | lymphoblasts': array(['CNN1', 'TMSB4X', 'MAPK1', 'GAL', 'ETS2', 'AIF1', 'SH3BGRL3',\n",
      "       'HBG2', 'CTSL', 'GMFG', 'AC079466.1', 'ZFP36L1', 'PRSS57',\n",
      "       'S100A11', 'RPL3', 'ARHGDIB', 'ISG15', 'HBG1', 'MDK', 'COTL1'],\n",
      "      dtype='<U10'), 'GLB1L2 | lymphoblasts': array(['TIMP1', 'CTNNB1', 'APOE', 'GAL', 'GLB1L2', 'TUBA1B', 'CKS1B',\n",
      "       'SET', 'YBX1', 'COL18A1', 'NEAT1', 'MYL4', 'UBE2S', 'AC079466.1',\n",
      "       'MT-ATP6', 'NCL', 'PSMB1', 'VIM', 'S100A13', 'RPL3'], dtype='<U10')}\n"
     ]
    }
   ],
   "source": [
    "# We obtain the differentially expressed genes (DEGs) for each perturbation; \n",
    "# Here we chose the top 20 non-zero expression DEGs for each perturbation from the adata.uns['rank_genes_groups_cov_all']\n",
    "print( dict(list(pert_data.adata.uns['top_non_zero_de_20'].items())[:5]) )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad4379d5",
   "metadata": {
    "papermill": {
     "duration": 0.151611,
     "end_time": "2025-06-18T07:15:57.184012",
     "exception": false,
     "start_time": "2025-06-18T07:15:57.032401",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Initialize the GEARS model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9780459f",
   "metadata": {
    "papermill": {
     "duration": 0.148502,
     "end_time": "2025-06-18T07:15:57.480299",
     "exception": false,
     "start_time": "2025-06-18T07:15:57.331797",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Set GEARS model parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4f617f79",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:15:57.781533Z",
     "iopub.status.busy": "2025-06-18T07:15:57.781251Z",
     "iopub.status.idle": "2025-06-18T07:15:58.883176Z",
     "shell.execute_reply": "2025-06-18T07:15:58.882370Z"
    },
    "papermill": {
     "duration": 1.254845,
     "end_time": "2025-06-18T07:15:58.884671",
     "exception": false,
     "start_time": "2025-06-18T07:15:57.629826",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# - init gears model\n",
    "pert_data.adata_split.X = pert_data.adata_split.X.toarray()\n",
    "gears_model = GEARS(pert_data, device = 'cuda:0', \n",
    "                        weight_bias_track = False, \n",
    "                        proj_name = 'pertnet', \n",
    "                        exp_name = 'pertnet')\n",
    "\n",
    "# - set model configuration\n",
    "gears_model.config = {'hidden_size': 64,\n",
    "                'num_go_gnn_layers' : 1, \n",
    "                'num_gene_gnn_layers' : 1,\n",
    "                'decoder_hidden_size' : 16,\n",
    "                'num_similar_genes_go_graph' : 20,\n",
    "                'num_similar_genes_co_express_graph' : 20,\n",
    "                'coexpress_threshold': 0.4,\n",
    "                'uncertainty' : False, \n",
    "                'uncertainty_reg' : 1,\n",
    "                'direction_lambda' : 1e-1,\n",
    "                'G_go': None,\n",
    "                'G_go_weight': None,\n",
    "                'G_coexpress': None,\n",
    "                'G_coexpress_weight': None,\n",
    "                'device': gears_model.device,\n",
    "                'num_genes': gears_model.num_genes,\n",
    "                'num_perts': gears_model.num_perts,\n",
    "                'no_perturb': False\n",
    "                }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff4ab82",
   "metadata": {
    "papermill": {
     "duration": 0.150238,
     "end_time": "2025-06-18T07:15:59.231982",
     "exception": false,
     "start_time": "2025-06-18T07:15:59.081744",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Construct co-expression graph and go graph. These two graphs are used to build the graph neural network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e86a268",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:15:59.530212Z",
     "iopub.status.busy": "2025-06-18T07:15:59.529812Z",
     "iopub.status.idle": "2025-06-18T07:16:23.588111Z",
     "shell.execute_reply": "2025-06-18T07:16:23.587547Z"
    },
    "papermill": {
     "duration": 24.209549,
     "end_time": "2025-06-18T07:16:23.589573",
     "exception": false,
     "start_time": "2025-06-18T07:15:59.380024",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found local copy...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this is lichen version 2!\n"
     ]
    }
   ],
   "source": [
    "# - Set the gene co expression network (green graph)\n",
    "if gears_model.config['G_coexpress'] is None:  # If the co-expression graph is not already set\n",
    "    ## calculating co expression similarity graph\n",
    "    edge_list = get_similarity_network(         # Compute the gene co-expression similarity network\n",
    "        network_type='co-express',              # Specify the network type as co-expression\n",
    "        adata=gears_model.adata,                # Pass the AnnData object with gene expression data\n",
    "        threshold=gears_model.config['coexpress_threshold'],  # Threshold for edge creation in the graph\n",
    "        k=gears_model.config['num_similar_genes_co_express_graph'],  # Number of similar genes to connect\n",
    "        data_path=gears_model.data_path,        # Path to the data directory\n",
    "        data_name=gears_model.dataset_name,     # Name of the dataset\n",
    "        split=gears_model.split, seed=gears_model.seed,  # Data split and random seed for reproducibility\n",
    "        train_gene_set_size=gears_model.train_gene_set_size,  # Size of the training gene set\n",
    "        set2conditions=gears_model.set2conditions           # Mapping from gene sets to experimental conditions\n",
    "    )\n",
    "\n",
    "    sim_network = GeneSimNetwork(               # Create a GeneSimNetwork object from the edge list\n",
    "        edge_list,                             # The computed edge list\n",
    "        gears_model.gene_list,                 # List of genes (nodes) in the graph\n",
    "        node_map=gears_model.node_map          # Mapping from gene names to node indices\n",
    "    )\n",
    "    gears_model.config['G_coexpress'] = sim_network.edge_index      # Store the edge indices in the config\n",
    "    gears_model.config['G_coexpress_weight'] = sim_network.edge_weight  # Store the edge weights in the config\n",
    "\n",
    "# - Set the gene ontology network (red graph)\n",
    "if gears_model.config['G_go'] is None:  # If the gene ontology (GO) graph is not already set\n",
    "    ## calculating gene ontology similarity graph\n",
    "    edge_list = get_similarity_network(         # Compute the gene ontology similarity network\n",
    "        network_type='go',                      # Specify the network type as gene ontology\n",
    "        adata=gears_model.adata,                # Pass the AnnData object\n",
    "        threshold=gears_model.config['coexpress_threshold'],  # Threshold for edge creation\n",
    "        k=gears_model.config['num_similar_genes_co_express_graph'],  # Number of similar genes to connect\n",
    "        pert_list=gears_model.pert_list,        # List of perturbations (for GO graph)\n",
    "        data_path=gears_model.data_path,        # Path to the data directory\n",
    "        data_name=gears_model.dataset_name,     # Name of the dataset\n",
    "        split=gears_model.split, seed=gears_model.seed,  # Data split and random seed\n",
    "        train_gene_set_size=gears_model.train_gene_set_size,  # Size of the training gene set\n",
    "        set2conditions=gears_model.set2conditions,            # Mapping from gene sets to conditions\n",
    "        default_pert_graph=gears_model.default_pert_graph     # Default perturbation graph (if any)\n",
    "    )\n",
    "\n",
    "    sim_network = GeneSimNetwork(               # Create a GeneSimNetwork object for the GO graph\n",
    "        edge_list,                             # The computed edge list\n",
    "        gears_model.pert_list,                 # List of perturbations (nodes in GO graph)\n",
    "        node_map=gears_model.node_map_pert     # Mapping from perturbation names to node indices\n",
    "    )\n",
    "    gears_model.config['G_go'] = sim_network.edge_index         # Store the edge indices in the config\n",
    "    gears_model.config['G_go_weight'] = sim_network.edge_weight # Store the edge weights in the config\n",
    "\n",
    "# - finally obtain the model\n",
    "gears_model.model = GEARS_Model(gears_model.config).to(gears_model.device)  # Instantiate and move the model to the device (e.g., GPU)\n",
    "gears_model.best_model = deepcopy(gears_model.model)                        # Save a copy as the best model so far"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9df12a73",
   "metadata": {
    "papermill": {
     "duration": 0.149884,
     "end_time": "2025-06-18T07:16:23.890393",
     "exception": false,
     "start_time": "2025-06-18T07:16:23.740509",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ac4202c",
   "metadata": {
    "papermill": {
     "duration": 0.199002,
     "end_time": "2025-06-18T07:16:24.237929",
     "exception": false,
     "start_time": "2025-06-18T07:16:24.038927",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Here we use a simplified pseudocode to demonstrate the core part of GEARS model. For the details of each line of code, you can refer to the source code.\n",
    "\n",
    "**Get Base Gene Embeddings**  \n",
    "In the following loop we use co-expression graph to obtain the final gene embeddings. (Green part in the figure above)\n",
    "<div style=\"background:#f9f9f9; border:1px solid #ccc; border-radius:6px; padding:12px; color:#333; font-family:monospace;\">\n",
    "<pre><code>\n",
    "base_emb = self.gene_emb(self.num_genes)\n",
    "pos_emb = self.emb_pos(self.num_genes)\n",
    "for idx, gnn_layer in enumerate(self.gnn_layers_exp):\n",
    "    pos_emb = gnn_layer(pos_emb, self.G_coexpress, self.G_coexpress_weight)\n",
    "base_emb = base_emb + pos_emb\n",
    "</code></pre>\n",
    "</div>\n",
    "Note that the gene embedding here can be directly replaced with LCM gene embedding (Such as scFoundation's embedding). these LCM embedding are pre-trained on huge amount of data and have richer and more comprehensive cellular information, which can enhance the performance in the Perturbation task.\n",
    "\n",
    "**Get Perturbation Embeddings**\n",
    "<div style=\"background:#f9f9f9; border:1px solid #ccc; border-radius:6px; padding:12px; color:#333; font-family:monospace;\">\n",
    "<pre><code>\n",
    "pert_global_emb = self.pert_emb(self.num_perts)\n",
    "</code></pre>\n",
    "</div>     \n",
    "\n",
    "In the following loop we use GO graph to obtain the final perturbation embeddings. (Red part in the figure above)\n",
    "<div style=\"background:#f9f9f9; border:1px solid #ccc; border-radius:6px; padding:12px; color:#333; font-family:monospace;\">\n",
    "<pre><code>\n",
    "for idx, gnn_layer in enumerate(self.gnn_layers_go):\n",
    "    pert_global_emb = gnn_layer(pert_global_emb, self.G_sim, self.G_sim_weight)\n",
    "</code></pre>\n",
    "</div>\n",
    "\n",
    "**Add Global Perturbation Embedding to Each Gene in Each Cell in the Batch**  \n",
    "(Composition Operator in the figure above)  \n",
    "Select the perturbation embedding of the coresponding gene\n",
    "<div style=\"background:#f9f9f9; border:1px solid #ccc; border-radius:6px; padding:12px; color:#333; font-family:monospace;\">\n",
    "<pre><code>\n",
    "pert_track = {}\n",
    "for i, j in enumerate(pert_index[0]):\n",
    "    pert_track[j.item()] = pert_global_emb[pert_index[1][i]]\n",
    "</code></pre>\n",
    "</div> \n",
    "Add the selected perturbation embedding to the gene embedding\n",
    "<div style=\"background:#f9f9f9; border:1px solid #ccc; border-radius:6px; padding:12px; color:#333; font-family:monospace;\">\n",
    "<pre><code>\n",
    "emb_total = self.pert_fuse(torch.stack(list(pert_track.values())))\n",
    "for idx, j in enumerate(pert_track.keys()):\n",
    "    base_emb[j] = base_emb[j] + emb_total[idx]\n",
    "</code></pre>\n",
    "</div> \n",
    "\n",
    "**Finally Go Through A Fully Connected Network to Obtain the Predicted Expression**  \n",
    "(Blue part in the figure above)  \n",
    "<div style=\"background:#f9f9f9; border:1px solid #ccc; border-radius:6px; padding:12px; color:#333; font-family:monospace;\">\n",
    "<pre><code>\n",
    "predict_expression = self.fcn(base_emb)\n",
    "</code></pre>\n",
    "</div> "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ea9756c",
   "metadata": {
    "papermill": {
     "duration": 0.14896,
     "end_time": "2025-06-18T07:16:24.540388",
     "exception": false,
     "start_time": "2025-06-18T07:16:24.391428",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Here we do not specifically delve into the training process of the model. The detail training process can be directly referred to the __gears_model.train__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "80b30926",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:16:24.841453Z",
     "iopub.status.busy": "2025-06-18T07:16:24.840811Z",
     "iopub.status.idle": "2025-06-18T07:18:42.984868Z",
     "shell.execute_reply": "2025-06-18T07:18:42.984110Z"
    },
    "papermill": {
     "duration": 138.425446,
     "end_time": "2025-06-18T07:18:43.115544",
     "exception": false,
     "start_time": "2025-06-18T07:16:24.690098",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Start Training...\n",
      "Epoch 1 Step 1 Train Loss: 0.6396\n",
      "Epoch 1 Step 51 Train Loss: 0.7346\n",
      "Epoch 1 Step 101 Train Loss: 0.8164\n",
      "Epoch 1 Step 151 Train Loss: 0.7041\n",
      "Epoch 1 Step 201 Train Loss: 0.7487\n",
      "Epoch 1 Step 251 Train Loss: 0.6544\n",
      "Epoch 1 Step 301 Train Loss: 0.6141\n",
      "Epoch 1 Step 351 Train Loss: 0.5417\n",
      "Epoch 1 Step 401 Train Loss: 0.6686\n",
      "Epoch 1 Step 451 Train Loss: 0.6415\n",
      "Epoch 1 Step 501 Train Loss: 0.6213\n",
      "Epoch 1 Step 551 Train Loss: 0.6144\n",
      "Epoch 1 Step 601 Train Loss: 0.6536\n",
      "Epoch 1 Step 651 Train Loss: 0.6656\n",
      "Epoch 1 Step 701 Train Loss: 0.5710\n",
      "Epoch 1 Step 751 Train Loss: 0.7103\n",
      "Epoch 1 Step 801 Train Loss: 0.6331\n",
      "Epoch 1 Step 851 Train Loss: 0.8571\n",
      "Epoch 1 Step 901 Train Loss: 0.6516\n",
      "Epoch 1 Step 951 Train Loss: 0.6747\n",
      "Epoch 1 Step 1001 Train Loss: 0.7110\n",
      "Epoch 1 Step 1051 Train Loss: 0.6432\n",
      "Epoch 1 Step 1101 Train Loss: 0.5850\n",
      "Epoch 1 Step 1151 Train Loss: 0.6577\n",
      "Epoch 1 Step 1201 Train Loss: 0.6702\n",
      "Epoch 1 Step 1251 Train Loss: 0.6095\n",
      "Epoch 1 Step 1301 Train Loss: 0.5903\n",
      "Epoch 1 Step 1351 Train Loss: 0.7571\n",
      "Epoch 1 Step 1401 Train Loss: 0.8350\n",
      "Epoch 1 Step 1451 Train Loss: 0.7280\n",
      "Epoch 1 Step 1501 Train Loss: 0.6518\n",
      "Epoch 1 Step 1551 Train Loss: 0.5930\n",
      "Epoch 1 Step 1601 Train Loss: 0.6536\n",
      "Epoch 1 Step 1651 Train Loss: 0.6568\n",
      "Epoch 1 Step 1701 Train Loss: 0.6420\n",
      "Epoch 1 Step 1751 Train Loss: 0.7210\n",
      "Epoch 1 Step 1801 Train Loss: 0.6687\n",
      "Epoch 1 Step 1851 Train Loss: 0.7130\n",
      "Epoch 1 Step 1901 Train Loss: 0.6425\n",
      "Epoch 1 Step 1951 Train Loss: 0.6809\n",
      "Epoch 1 Step 2001 Train Loss: 0.7343\n",
      "Epoch 1 Step 2051 Train Loss: 0.7038\n",
      "Epoch 1: Train Overall MSE: 0.0058 Validation Overall MSE: 0.0072. \n",
      "Train Top 20 DE MSE: 0.0538 Validation Top 20 DE MSE: 0.1644. \n",
      "Done!\n",
      "Done! No test dataloader detected.\n"
     ]
    }
   ],
   "source": [
    "gears_model.train(epochs = 1, lr = 1e-4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23e3d076",
   "metadata": {
    "papermill": {
     "duration": 0.150211,
     "end_time": "2025-06-18T07:18:43.417822",
     "exception": false,
     "start_time": "2025-06-18T07:18:43.267611",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Evaluate the performance  \n",
    "Here we use the test set to evaluate the model's performance. All the perturbed genes in the test set are not seen in the training set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "33d4eeb5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:18:43.722021Z",
     "iopub.status.busy": "2025-06-18T07:18:43.721753Z",
     "iopub.status.idle": "2025-06-18T07:18:50.652100Z",
     "shell.execute_reply": "2025-06-18T07:18:50.651490Z"
    },
    "papermill": {
     "duration": 7.08354,
     "end_time": "2025-06-18T07:18:50.653550",
     "exception": false,
     "start_time": "2025-06-18T07:18:43.570010",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# - Model testing\n",
    "test_output = evaluate(testloader, gears_model.best_model,\n",
    "                    gears_model.config['uncertainty'], gears_model.device)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fe9d225",
   "metadata": {
    "papermill": {
     "duration": 0.153259,
     "end_time": "2025-06-18T07:18:50.964541",
     "exception": false,
     "start_time": "2025-06-18T07:18:50.811282",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "We can then calculate the evaluation metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "651abe74",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:18:51.316619Z",
     "iopub.status.busy": "2025-06-18T07:18:51.316336Z",
     "iopub.status.idle": "2025-06-18T07:18:53.047664Z",
     "shell.execute_reply": "2025-06-18T07:18:53.046920Z"
    },
    "papermill": {
     "duration": 1.933287,
     "end_time": "2025-06-18T07:18:53.049300",
     "exception": false,
     "start_time": "2025-06-18T07:18:51.116013",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "pert_metric = {}\n",
    "# here we take one perturbation as an example\n",
    "# this function will first prepare the data for metric calculation\n",
    "de_idx_map, ctrl, p_idx = prepare_for_metric(pert_data.adata, test_output, pert = np.unique(test_output['pert_cat'])[0],\n",
    "                                              most_variable_genes=None, p_thre_1=0.01, p_thre_2=0.1)\n",
    "# de_idx_map: the gene index of top20, top50 and top100 DEGs\n",
    "# ctrl: the control group cell for this perturbed gene\n",
    "# p_idx: the perturbed cell index for this perturbed gene"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e66b28e",
   "metadata": {
    "papermill": {
     "duration": 0.155443,
     "end_time": "2025-06-18T07:18:53.368668",
     "exception": false,
     "start_time": "2025-06-18T07:18:53.213225",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "We use the following metrics for evaluation\n",
    "\n",
    "**pearson correlation**: measures the linear correlation between predicted and true values  \n",
    "**mean squared error**: measures the average of the squares of the errors  \n",
    "**mean absolute error**: measures the average of the absolute errors  \n",
    "**change ratio**: measures the relative change between predicted and true values  \n",
    "**spearman correlation**: measures the monotonic relationship between predicted and true values  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fe9db4d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:18:53.673582Z",
     "iopub.status.busy": "2025-06-18T07:18:53.673305Z",
     "iopub.status.idle": "2025-06-18T07:18:53.723308Z",
     "shell.execute_reply": "2025-06-18T07:18:53.722713Z"
    },
    "papermill": {
     "duration": 0.203318,
     "end_time": "2025-06-18T07:18:53.724450",
     "exception": false,
     "start_time": "2025-06-18T07:18:53.521132",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define a dictionary mapping metric names to their corresponding functions\n",
    "metric2fct = {\n",
    "    'pearson': pearsonr,           # Pearson correlation coefficient\n",
    "    'mse': mse,                    # Mean squared error\n",
    "    'mae': mae,                    # Mean absolute error\n",
    "    'change_ratio': get_change_ratio,  # Custom function for change ratio\n",
    "    'spearman': spearmanr,         # Spearman correlation coefficient\n",
    "}\n",
    "name = 'DE_'  # Prefix for metric names in the results dictionary\n",
    "\n",
    "# Loop over each DEG set (e.g., 'top20', 'top50', 'top100')\n",
    "for prefix in list(de_idx_map.keys()):  # de_idx_map contains indices for each DEG set\n",
    "    de_idx = de_idx_map[prefix]         # Get the indices for the current DEG set\n",
    "    for m, fct in metric2fct.items():   # Loop over each metric and its function\n",
    "        if m == 'pearson' or m == 'spearman':\n",
    "            # Calculate the metric on delta expression (perturbed - control)\n",
    "            val = fct(\n",
    "                test_output['pred'][p_idx].mean(0)[de_idx] - ctrl[de_idx],   # Predicted delta\n",
    "                test_output['truth'][p_idx].mean(0)[de_idx] - ctrl[de_idx]   # True delta\n",
    "            )[0]\n",
    "            if np.isnan(val):  # Handle NaN results\n",
    "                val = 0\n",
    "            pert_metric[name + m + f'_delta_{prefix}'] = val  # Store delta metric\n",
    "\n",
    "            # Calculate the metric on raw predicted vs. true expression\n",
    "            val = fct(\n",
    "                test_output['pred'][p_idx].mean(0)[de_idx],    # Predicted expression\n",
    "                test_output['truth'][p_idx].mean(0)[de_idx]    # True expression\n",
    "            )[0]\n",
    "            if np.isnan(val):\n",
    "                val = 0\n",
    "            pert_metric[name + m + f'_{prefix}'] = val         # Store raw metric\n",
    "        elif m == 'change_ratio':\n",
    "            # Calculate the change ratio metric (custom function)\n",
    "            val = fct(\n",
    "                test_output['pred'][p_idx].mean(0)[de_idx],\n",
    "                test_output['truth'][p_idx].mean(0)[de_idx]\n",
    "            )\n",
    "            pert_metric[name + m + f'_{prefix}'] = val\n",
    "        else:\n",
    "            # For mse and mae, calculate on delta expression (perturbed - control)\n",
    "            val = fct(\n",
    "                test_output['pred'][p_idx].mean(0)[de_idx] - ctrl[de_idx],\n",
    "                test_output['truth'][p_idx].mean(0)[de_idx] - ctrl[de_idx]\n",
    "            )\n",
    "            pert_metric[name + m +"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1e8378ed",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-18T07:18:54.030952Z",
     "iopub.status.busy": "2025-06-18T07:18:54.030694Z",
     "iopub.status.idle": "2025-06-18T07:18:54.035832Z",
     "shell.execute_reply": "2025-06-18T07:18:54.035252Z"
    },
    "papermill": {
     "duration": 0.159306,
     "end_time": "2025-06-18T07:18:54.036983",
     "exception": false,
     "start_time": "2025-06-18T07:18:53.877677",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DE_pearson_delta_top20': 0.6932509,\n",
       " 'DE_pearson_top20': 0.9079151,\n",
       " 'DE_mse_top20': 0.29012805,\n",
       " 'DE_mae_top20': 0.45869684,\n",
       " 'DE_change_ratio_top20': 0.35481155,\n",
       " 'DE_spearman_delta_top20': 0.6962406015037594,\n",
       " 'DE_spearman_top20': 0.8736842105263156,\n",
       " 'DE_pearson_delta_top50': 0.6354232,\n",
       " 'DE_pearson_top50': 0.9486616,\n",
       " 'DE_mse_top50': 0.14973007,\n",
       " 'DE_mae_top50': 0.32067817,\n",
       " 'DE_change_ratio_top50': 0.31630465,\n",
       " 'DE_spearman_delta_top50': 0.6956542617046818,\n",
       " 'DE_spearman_top50': 0.892436974789916,\n",
       " 'DE_pearson_delta_top100': 0.54854935,\n",
       " 'DE_pearson_top100': 0.96458554,\n",
       " 'DE_mse_top100': 0.08678662,\n",
       " 'DE_mae_top100': 0.2287849,\n",
       " 'DE_change_ratio_top100': 0.33833167,\n",
       " 'DE_spearman_delta_top100': 0.5639603960396039,\n",
       " 'DE_spearman_top100': 0.9119711971197119,\n",
       " 'DE_pearson_delta_top200': 0.46559995,\n",
       " 'DE_pearson_top200': 0.9763876,\n",
       " 'DE_mse_top200': 0.045821756,\n",
       " 'DE_mae_top200': 0.13980067,\n",
       " 'DE_change_ratio_top200': 1.1450596,\n",
       " 'DE_spearman_delta_top200': 0.42882098767629945,\n",
       " 'DE_spearman_top200': 0.9610443385232474,\n",
       " 'DE_pearson_delta_DE_sig': 0.57420224,\n",
       " 'DE_pearson_DE_sig': 0.9592006,\n",
       " 'DE_mse_DE_sig': 0.10758097,\n",
       " 'DE_mae_DE_sig': 0.26323307,\n",
       " 'DE_change_ratio_DE_sig': 0.33999825,\n",
       " 'DE_spearman_delta_DE_sig': 0.6098584959344453,\n",
       " 'DE_spearman_DE_sig': 0.9135547996307489}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pert_metric"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "660a2052",
   "metadata": {},
   "source": [
    "### Only trained for 1 epoch so didn't get good results - can try running for more epochs\n",
    "Can also change the training dataset"
   ]
  }
 ],
 "metadata": {
  "kaggle": {
   "accelerator": "gpu",
   "dataSources": [
    {
     "datasetId": 7686605,
     "sourceId": 12202625,
     "sourceType": "datasetVersion"
    },
    {
     "isSourceIdPinned": false,
     "modelId": 376682,
     "modelInstanceId": 355372,
     "sourceId": 435716,
     "sourceType": "modelInstanceVersion"
    },
    {
     "isSourceIdPinned": false,
     "modelId": 378192,
     "modelInstanceId": 356881,
     "sourceId": 437491,
     "sourceType": "modelInstanceVersion"
    }
   ],
   "dockerImageVersionId": 31041,
   "isGpuEnabled": true,
   "isInternetEnabled": true,
   "language": "python",
   "sourceType": "notebook"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 393.386851,
   "end_time": "2025-06-18T07:18:57.877546",
   "environment_variables": {},
   "exception": null,
   "input_path": "__notebook__.ipynb",
   "output_path": "__notebook__.ipynb",
   "parameters": {},
   "start_time": "2025-06-18T07:12:24.490695",
   "version": "2.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
