---
title: "Mendelian Randomization Tutorial: CRP, IL-6, and CHD"
author: "Tayyaba Alvi"
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
library(knitr)
library(TwoSampleMR)
library(ieugwasr)
library(dplyr)
library(VariantAnnotation)
library(MRInstruments)
library(gwasglue)
library(ggplot2)
```

# Introduction

This tutorial demonstrates how to perform a Mendelian Randomization (MR) analysis using multiple data sources. We explore the causal relationship between C-reactive protein (CRP) and coronary heart disease (CHD), followed by a deeper look at interleukin-6 (IL-6), a potential upstream regulator of CRP. The results highlight the contrast between observational and MR-based evidence and suggest that IL-6 signalling may play a more direct causal role in CHD risk.

# 1. Accessing GWAS Data from Different Sources

## 1.1 From Local VCF Files Using `gwasvcf_to_TwoSampleMR`
```{r  message=FALSE, warning=FALSE}
vcf_crp <- VariantAnnotation::readVcf("data/ieugwas/CRP_ebi-a-GCST005067_sub.vcf.gz")
crp_exposure <- gwasvcf_to_TwoSampleMR(vcf_crp, "exposure")
crp_exposure <- crp_exposure %>% filter(pval.exposure < 5*10^-8)

vcf_chd <- VariantAnnotation::readVcf("data/ieugwas/CHD_ebi-a-GCST000998_sub.vcf.gz")
chd_outcome <- gwasvcf_to_TwoSampleMR(vcf_chd, "outcome")

harm_data <- harmonise_data(exposure_dat = crp_exposure, outcome_dat = chd_outcome)
```

## 1.2 From MRBase API Using `MRInstruments`
```{r  message=FALSE, warning=FALSE}
vignette("MRBase")
data(gwas_catalog)
exposure_gwas <- subset(gwas_catalog, grepl("GCST007615", STUDY.ACCESSION) &
                          Phenotype_simple == "C-reactive protein levels")
exposure_gwas <- exposure_gwas[exposure_gwas$pval < 5*10^-8, ]
exposure_data <- format_data(exposure_gwas)
head(exposure_data)

# The new update requires authentication through accsess token. More details at https://mrcieu.github.io/ieugwasr/articles/guide.html#authentication
# Check available outcomes
#ao <- available_outcomes()
#head(ao)

# Extract CHD outcome data
#chd_outcome_api <- extract_outcome_data(
#  snps = exposure_data$SNP,
#  outcomes = 'ieu-a-7'
#)
```

## 1.3 From Local TSV Files Using `read_exposure_data`
```{r  message=FALSE, warning=FALSE}
exposure <- read_exposure_data(
  filename = "data/gwas_catalogue/CRP_gwas_catalogue.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
exposure$exposure <- "CRP"
exposure <- exposure %>% filter(pval.exposure < 5*10^-8)

outcome <- read_outcome_data(
  snps = exposure$SNP,
  filename = "data/gwas_catalogue/CHD_gwas_catalogue.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
outcome$outcome <- "CHD"
```
# Overview

This part investigates the causal cascade of IL-6 signaling,  C-reactive protein (CRP), and coronary heart disease (CHD) using Mendelian Randomization (MR). We'll test whether genetic proxies for IL-6, IL6R, and CRP are causally associated with each other and with CHD.

**Summary of the cascade:**

1. **IL6R → CRP**: Evidence supports a strong causal effect, consistent with IL6R as an upstream driver of inflammation.
2. **CRP → CHD**: Evidence **does not** support a causal effect, suggesting CRP may be a **marker** rather than a driver of CHD.
3. **IL-6 → CHD**: Results suggest a potential causal role for IL-6 in CHD risk, but pleiotropy and heterogeneity must be considered.
4. **IL6R → CHD**: Provides a druggable link; may support IL6R-targeting interventions like tocilizumab.

# Harmonize and clump
```{r  message=FALSE, warning=FALSE}
dat <- harmonise_data(exposure_dat = exposure, outcome_dat = outcome)
clump <- ld_clump(
  dplyr::tibble(rsid=dat$SNP, pval=dat$pval.exposure, id=dat$exposure),
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p = 0.99,
  pop = "EUR",
  opengwas_jwt = get_opengwas_jwt(),
  bfile = NULL,
  plink_bin = NULL
)
#the ld_clump function uses API to clump provided SNPs using LD information from 1kg data. If this fails to run then we can also use local plink binaries for LD reference. link: https://mrcieu.github.io/gwasglue/index.html
#clump <- ld_clump(
#  dplyr::tibble(rsid=dat$SNP, pval=dat$pval.exposure, id=dat$exposure),
#  pop = "EUR",
#  plink_bin = genetics.binaRies::get_plink_binary(),
#  bfile = "data/1kg.v3/EUR"
#)
clump <- readRDS("data/clump_crp.rds")
harm_clump <- merge(dat, clump, by.x = c("SNP", "pval.exposure", "exposure"), by.y = c("rsid", "pval", "id")) %>% unique()
```

# 1. MR Analysis of CRP on CHD

Let's test whether CRP levels have a causal effect on coronary heart disease.
---
```{r}
res <- mr(harm_clump, method_list = c("mr_egger_regression", "mr_ivw"))
res

mr_heterogeneity(harm_clump)
mr_pleiotropy_test(harm_clump)

# Plots
p1 <- mr_scatter_plot(res, harm_clump)
p1

res_single <- mr_singlesnp(harm_clump)
p2 <- mr_forest_plot(res_single)
p2

res_loo <- mr_leaveoneout(harm_clump)
p3 <- mr_leaveoneout_plot(res_loo)
p3

# Generate report
#mr_report(harm_clump, output_path = "figure/")
```

**Exercise:**
- Do we observe a causal effect of CRP on CHD?
- Are the MR assumptions violated based on pleiotropy and heterogeneity tests?

# 2. MR Analysis of IL-6 on CHD

Let's test whether IL-6 levels are causally associated with CHD risk.

```{r  message=FALSE, warning=FALSE}
exposure <- read_exposure_data(
  filename = "data/gwas_catalogue/IL6-gwas_catalogue_sub.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
exposure$exposure <- "IL-6"
exposure <- exposure %>% filter(pval.exposure < 5*10^-8)

outcome <- read_outcome_data(
  snps = exposure$SNP,
  filename = "data/gwas_catalogue/CHD_gwas_catalogue.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
outcome$outcome <- "CHD"

il6_data <- harmonise_data(exposure_dat = exposure, outcome_dat = outcome)
clump <- ld_clump(
  dplyr::tibble(rsid=il6_data$SNP, pval=il6_data$pval.exposure,exposure=il6_data$exposure),
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p = 0.99,
  pop = "EUR",
  opengwas_jwt = get_opengwas_jwt(),
  bfile = NULL,
  plink_bin = NULL
)
clump <- readRDS("data/il6_clump.rds")
harm_clump_il6 <- merge(il6_data, clump, by.x = c("SNP", "pval.exposure", "exposure"), by.y = c("rsid", "pval", "exposure")) %>% unique()
#filter for variant rs2228145
res_il6 <- mr_singlesnp(harm_clump_il6)
res_il6 <- res_il6[1,] %>%
  mutate(
    ci.lower = b - 1.96 * se,
    ci.upper = b + 1.96 * se
  )
p2 <- ggplot(res_il6, aes(x = b, y = SNP)) +
  geom_point(size = 3, color = "steelblue") +
  geom_errorbar(aes(xmin = ci.lower, xmax = ci.upper), width = 0.1, color = "steelblue") +
  geom_hline(yintercept = 0, linetype = "dashed") +
  labs(
    title = paste0("Causal Effect of ", res_il6$exposure, " on ", res_il6$outcome),
    x = "Causal Estimate",
    y = "SNP"
  ) +
  theme_minimal(base_size = 14)
p2

#Strong evidence suggesting rs2228145 of the Interleukin 6 receptor (IL-6R Asp358Ala variant) is associated with a 2-fold increase in soluble IL-6R (sIL-6R) serum levels resulting in reduced IL-6-induced C-reactive protein (CRP) production and a reduced risk for coronary heart disease. 
```

**Exercise:**
- Is IL-6 likely to be causally associated with CHD?


# 3. MR Analysis of IL6R on CRP

Let's test whether IL6R levels influence CRP levels.
```{r  message=FALSE, warning=FALSE}
exposure <- read_exposure_data(
  filename = "data/gwas_catalogue/IL6R-gwas_catalogue.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
exposure$exposure <- "IL6R"
exposure <- exposure %>% filter(pval.exposure < 5*10^-8)

outcome <- read_outcome_data(
  snps = exposure$SNP,
  filename = "data/gwas_catalogue/CRP_gwas_catalogue.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
outcome$outcome <- "CRP"

il6R_data <- harmonise_data(exposure_dat = exposure, outcome_dat = outcome)
clump <- ld_clump(
  dplyr::tibble(rsid=il6_data$SNP, pval=il6_data$pval.exposure,exposure=il6_data$exposure),
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p = 0.99,
  pop = "EUR",
  opengwas_jwt = get_opengwas_jwt(),
  bfile = NULL,
  plink_bin = NULL
)
clump <- readRDS("data/il6R_crp_clump.rds")
harm_clump_il6R_crp <- merge(il6R_data, clump, by.x = c("SNP", "pval.exposure", "exposure"), by.y = c("rsid", "pval", "exposure")) %>% unique()

res_il6R_crp <- mr(harm_clump_il6R_crp, method_list = c("mr_egger_regression", "mr_ivw"))
res_il6R_crp
mr_heterogeneity(harm_clump_il6R_crp)
mr_pleiotropy_test(harm_clump_il6R_crp)

p1_il6R <- mr_scatter_plot(res_il6R_crp, harm_clump_il6R_crp)
p1_il6R

res_single_il6R <- mr_singlesnp(harm_clump_il6R_crp)
p2_il6R <- mr_forest_plot(res_single_il6R)
p2_il6R

```

**Exercise:**
- Is there evidence that IL6R variants causally affect CRP levels?
- Does this support IL6R as an upstream regulator of inflammation?

# 4. MR Analysis of IL6R on CHD

Finally, let's test whether IL6R levels have a causal effect on CHD.
```{r  message=FALSE, warning=FALSE}
exposure <- read_exposure_data(
  filename = "data/gwas_catalogue/IL6R-gwas_catalogue.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
exposure$exposure <- "IL6R"
exposure <- exposure %>% filter(pval.exposure < 5*10^-8)

outcome <- read_outcome_data(
  snps = exposure$SNP,
  filename = "data/gwas_catalogue/CHD_gwas_catalogue.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
outcome$outcome <- "CHD"

il6R_data <- harmonise_data(exposure_dat = exposure, outcome_dat = outcome)
clump <- ld_clump(
  dplyr::tibble(rsid=il6R_data$SNP, pval=il6R_data$pval.exposure,exposure=il6R_data$exposure),
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p = 0.99,
  pop = "EUR",
  opengwas_jwt = get_opengwas_jwt(),
  bfile = NULL,
  plink_bin = NULL
)
clump <- readRDS("data/clump_il6R_chd.rds")
harm_clump_il6R <- merge(il6R_data, clump, by.x = c("SNP", "pval.exposure", "exposure"), by.y = c("rsid", "pval", "exposure")) %>% unique()

res_il6R <- mr(harm_clump_il6R, method_list = c("mr_egger_regression", "mr_ivw"))
res_il6R
mr_heterogeneity(harm_clump_il6R)
mr_pleiotropy_test(harm_clump_il6R)

p1_il6R <- mr_scatter_plot(res_il6R, harm_clump_il6R)
p1_il6R

res_single_il6R <- mr_singlesnp(harm_clump_il6R)
p2_il6R <- mr_forest_plot(res_single_il6R)
p2_il6R

```

**Exercise:**
- Do IL6R variants predict CHD risk?
- Would this support therapeutic targeting of IL6R (e.g., using tocilizumab)?


# 5: Reverse Causality: CHD -> IL6R

```{r  message=FALSE, warning=FALSE}
exposure <- read_exposure_data(
  filename = "data/gwas_catalogue/CHD_gwas_catalogue.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
exposure$exposure <- "CHD"
exposure <- exposure %>% filter(pval.exposure < 5*10^-8)

outcome <- read_outcome_data(
  snps = exposure$SNP,
  filename = "data/gwas_catalogue/IL6R-gwas_catalogue.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
outcome$outcome <- "IL6-R"
dat <- harmonise_data(exposure_dat = exposure, outcome_dat = outcome)
clump <- ld_clump(
  dplyr::tibble(rsid=dat$SNP, pval=dat$pval.exposure, id=dat$exposure),
  pop = "EUR",
  plink_bin = genetics.binaRies::get_plink_binary(),
  bfile = "data/1kg.v3/EUR"
)

harm_clump <- merge(dat, clump, by.x = c("SNP", "pval.exposure", "exposure"), by.y = c("rsid", "pval", "id")) %>% unique()
res <- mr(harm_clump, method_list = c("mr_egger_regression", "mr_ivw"))
res
p1 <- mr_scatter_plot(res, harm_clump)
p1
mr_heterogeneity(harm_clump)
mr_pleiotropy_test(harm_clump)
```

These results suggest that **targeting IL6R** may influence CHD risk through inflammatory pathways without CRP being a causal intermediary. This has direct implications for drug repurposing and therapeutic targeting.

# Add-on: Reporting MR Estimates as Odds Ratios

MR effect estimates are often reported as odds ratios (OR) with confidence intervals (CI). 

```{r  message=FALSE, warning=FALSE}
# Assuming `mr_results` contains columns: method, b, se

mr_results_odds <- res_il6R %>%
  mutate(
    OR = exp(b),
    CI_lower = exp(b - 1.96 * se),
    CI_upper = exp(b + 1.96 * se)
  )

# Plotting Odds Ratios with ggplot2
ggplot(mr_results_odds, aes(x = OR, y = method)) +
  geom_point() +
  geom_errorbar(aes(xmin = CI_lower, xmax = CI_upper), width = 0) +
  geom_vline(xintercept = 1, linetype = "dashed") +
  xlab("Odds Ratio") +
  ylab("MR Method") +
  theme_bw()

# This plot helps visualize the causal effect of LDL on CAD across different MR methods and whether the 95% confidence intervals exclude the null (OR = 1).

```

# Conclusion

This tutorial demonstrated how to conduct MR using multiple data formats (VCF, TSV, API) and data harmonization pipelines. The CRP–CHD analysis showed no evidence of a causal effect, while IL-6—upstream of CRP—demonstrated a potential causal effect on CHD, supporting the hypothesis that IL-6 may be a better intervention target. Reverse causality check confirms that the uni-directional relationship between iL6 -> CHD.

